ESBMC fails to compile on LLVM-17 and LLVM-18

When trying to compile ESBMC with recent versions of llvm:

```
/Users/user/Repos/esbmc/src/clang-c-frontend/AST/build_ast.cpp:17:10: fatal error: 'llvm/Support/Host.h' file not found
#include <llvm/Support/Host.h>
```

And checking on my environment:

```
$ ls llvm@16/include/llvm/Support | grep Host
Host.h

$ ls llvm@17/include/llvm/Support | grep Host

$ ls llvm@18/include/llvm/Support | grep Host

```

Guess they changed the header location (or removed it).

Works for llvm-17 for me:
```
$ ll /usr/lib/llvm/17/include/llvm/Support/Host.h 
-rw-r--r-- 1 root root 749 Nov 28 09:52 /usr/lib/llvm/17/include/llvm/Support/Host.h
```
llvm-18 fails:
```
/home/kane/scorch/esbmc/src/clang-cpp-frontend/clang_cpp_convert_vft.cpp: In member function 'bool clang_cpp_convertert::annotate_virtual_overriding_methods(const clang::CXXMethodDecl&, struct_union_typet::componentt&)':
/home/kane/scorch/esbmc/src/clang-cpp-frontend/clang_cpp_convert_vft.cpp:112:10: error: 'const class clang::CXXMethodDecl' has no member named 'isPure'
  112 |   if (md.isPure())
      |          ^~~~~~
```

--- a/src/clang-c-frontend/AST/build_ast.cpp
+++ b/src/clang-c-frontend/AST/build_ast.cpp
@@ -14,7 +14,11 @@ CC_DIAGNOSTIC_IGNORE_LLVM_CHECKS()
 #include <clang/Lex/PreprocessorOptions.h>
 #include <clang/Tooling/Tooling.h>
 #include <llvm/Option/ArgList.h>
+#if CLANG_VERSION_MAJOR < 16
 #include <llvm/Support/Host.h>
+#else
+#include <llvm/TargetParser/Host.h>
+#endif
 #include <llvm/Support/Path.h>
 CC_DIAGNOSTIC_POP()
 

--- a/src/clang-cpp-frontend/clang_cpp_convert_vft.cpp
+++ b/src/clang-cpp-frontend/clang_cpp_convert_vft.cpp
@@ -9,6 +9,7 @@
 // Remove warnings from Clang headers
 CC_DIAGNOSTIC_PUSH()
 CC_DIAGNOSTIC_IGNORE_LLVM_CHECKS()
+#include <clang/Basic/Version.inc>
 #include <clang/AST/Attr.h>
 #include <clang/AST/DeclCXX.h>
 #include <clang/AST/DeclFriend.h>
@@ -97,6 +98,15 @@ bool clang_cpp_convertert::get_struct_class_virtual_methods(
   return false;
 }
 
+static bool is_pure_virtual(const clang::CXXMethodDecl &md)
+{
+#if CLANG_VERSION_MAJOR < 18
+  return md.isPure();
+#else
+  return md.isPureVirtual();
+#endif
+}
+
 bool clang_cpp_convertert::annotate_virtual_overriding_methods(
   const clang::CXXMethodDecl &md,
   struct_typet::componentt &comp)
@@ -109,7 +119,7 @@ bool clang_cpp_convertert::annotate_virtual_overriding_methods(
   comp.set("is_virtual", true);
   comp.set("virtual_name", method_name);
 
-  if (md.isPure())
+  if (is_pure_virtual(md))
     comp.set("is_pure_virtual", true);
 
   return false;

